### DISCUSSION TITLE
NixOps Architecture

### DISCUSSION THREAD

**Discussion by DasenB** [nixops4/nixops4#13] [General]
I have read the documents in the current pull requests but I still struggle a bit to understand the exact structure. My current understanding ist the following:

A user that wants to deploy to a machine describes in a flake the desired system configuration and the deployment target. The system ist described with a standard nix config. The deployment is described with a nix expression.

The expressions generate the actual commands that are executed to create/modify/delete/update/… a deployment. The development of these expressions is related to the nixops but project not a part of it.

NixOps takes the nix expression that the user supplies, evaluates it and executes the resulting commands as needed (e.g change a server over ssh, run docker commands, create a aws instance…)

Do I understand this correctly?

**Comment by scottbot95** [comment]
That is more-or-less my understanding after reading the documents as well. 

One thing that seemed missing from the docs to me is resource deletion. From #7 it sounds like `nixops` will invoke the resource provider, then simply pass it the properties needed to create a resource at which point the resource provider will attempt to create the resource. Surely NixOps needs to support deletion of resources as well and it doesn't seem obvious to me how that fits into the current design.

**Reply by aanderse** [reply]
presumably these provisioning actions are passed off to `opentofu`

**Reply by scottbot95** [reply]
Hmmm that brings up another good question actually. How does OpenTofu fit in more generally? It's mentioned in the top level readme, but doesn't really seem to be mentioned in the ADRs

**Reply by aanderse** [reply]
if you can [write terraform code in nix](https://github.com/terranix/terranix) you can do some cool things like create high level interfaces on top of the low level `terraform`... like what NixOS does on top of a linux system

though from the notes written about sustainability it sounds like a more direct approach, more in line with [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-2-balancing-module-option-count) concepts, is probably the idea

**Reply by roberth** [reply]
Thanks for bringing up deletions. How deletions can be done differs between stateful and stateless deployments, so I think that should be added to https://github.com/nixops4/nixops4/pull/8/files

**Reply by roberth** [reply]
I think the initial approach for OpenTofu integration is to NixOps resources that invoke OpenTofu commands on generated OpenTofu declarations.
This might be inefficient, without reusing the provider process, so a next step could be to allow NixOps resources to start such a process and reuse it.
It don't think the integration needs to be in the core, but I wouldn't rule it out entirely.

**Comment by roberth** [comment]
@DasenB sounds about right. I haven't spent many words on the aspects that remain similar to NixOps 1 and 2, but you've filled that in quite nicely.
A high level description like that will make its way into the user docs.

**Comment by nh2** [comment]
I've added some more practical requirements that the architecture should cover here:

https://github.com/NixOS/nixops/issues/1574#issuecomment-2192917900
